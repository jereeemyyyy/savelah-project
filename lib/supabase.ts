import { AppState } from 'react-native'
import 'react-native-url-polyfill/auto'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = 'https://cjcldaiiqmedjmojjzfz.supabase.co'
const supabaseAnonKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNqY2xkYWlpcW1lZGptb2pqemZ6Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MTY3OTE2NDksImV4cCI6MjAzMjM2NzY0OX0.6z2M2mG_qmSPPwujMWYpN4KkLgiU0mWjGdsCdVZvlNw'

export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
})

// Tells Supabase Auth to continuously refresh the session automatically
// if the app is in the foreground. When this is added, you will continue
// to receive `onAuthStateChange` events with the `TOKEN_REFRESHED` or
// `SIGNED_OUT` event if the user's session is terminated. This should
// only be registered once.
AppState.addEventListener('change', (state) => {
  if (state === 'active') {
    supabase.auth.startAutoRefresh()
  } else {
    supabase.auth.stopAutoRefresh()
  }
})