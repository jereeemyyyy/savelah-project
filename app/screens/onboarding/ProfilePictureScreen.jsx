import { View, Text, TouchableOpacity, Image, Alert } from 'react-native';
import { useNavigation, useRoute } from "@react-navigation/native";
import BackButton from '../../components/BackButton';
import * as ImagePicker from 'expo-image-picker';
import { supabase } from '../../../lib/supabase';
import { useState } from 'react';
import { MaterialIcons } from '@expo/vector-icons';

export default function ProfilePictureScreen() {
    const navigation = useNavigation();
    const route = useRoute();
    const { username, userId } = route.params;
    const [profilePicture, setProfilePicture] = useState(null);

    const handleNext = () => {
        navigation.navigate('BankDetails', { username, userId });
    };

    const pickProfilePicture = async () => {
        let result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.All,
            allowsEditing: true,
            aspect: [4, 3],
            quality: 1,
        });

        if (!result.canceled) {
            const source = result.assets[0];
            const fileExt = source.uri.split('.').pop();
            const fileName = `${userId}.${fileExt}`;
            const filePath = `profile-pictures/${fileName}_${Date.now()}`;

            const formData = new FormData();
            formData.append('file', {
                uri: source.uri,
                name: fileName,
                type: `image/${fileExt}`
            });

            let { error: uploadError } = await supabase.storage
                .from('profile-pictures')
                .upload(filePath, formData);

            if (uploadError) {
                console.error('Error uploading image:', uploadError.message);
                Alert.alert('Error', 'Failed to upload profile picture');
                return;
            }

            const { data: imageUrl } = await supabase.storage
                .from('profile-pictures')
                .getPublicUrl(filePath);

            const { error: updateError } = await supabase
                .from('profiles')
                .update({ profile_picture: imageUrl.publicUrl })
                .eq('id', userId);

            if (updateError) {
                console.error('Error updating profile picture:', updateError.message);
            } else {
                setProfilePicture(imageUrl.publicUrl);
            }
        }
    }; 

    return (
        <View className="flex-1 bg-gray-800">
            <BackButton />
            <Text className="text-4xl font-bold text-white m-4">Add a Profile Picture</Text>
            <Text className="text-lg text-gray-400 mx-4 my-2">Customize your profile by adding a Profile Picture!</Text>
            <TouchableOpacity className="items-center my-4" onPress={pickProfilePicture}>
                {profilePicture ? (
                    <Image source={{ uri: profilePicture }} className="w-36 h-36 rounded-full" />
                ) : (
                    <View className="w-36 h-36 bg-gray-200 rounded-full items-center justify-center">
                        <MaterialIcons name="add-a-photo" size={24} color="black"/>
                    </View>
                )}
            </TouchableOpacity>

            <View className="flex-1 items-center m-4">
                <TouchableOpacity className="py-3 bg-indigo-500 rounded mb-4 w-20 items-center" onPress={handleNext}>
                    <Text className="text-white">Next</Text>
                </TouchableOpacity>
                <TouchableOpacity className="py-3 bg-gray-500 rounded w-20 items-center" onPress={handleNext}>
                    <Text className="text-white">Skip</Text>
                </TouchableOpacity>
            </View>
        </View>
    );
}
