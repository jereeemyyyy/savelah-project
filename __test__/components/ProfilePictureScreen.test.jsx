import React from 'react';
import renderer from 'react-test-renderer';
import ProfilePictureScreen from '../../app/screens/onboarding/ProfilePictureScreen';
import { useNavigation, useRoute } from '@react-navigation/native';
import * as ImagePicker from 'expo-image-picker';
import { supabase } from '../../lib/supabase';

// Mock navigation and route
jest.mock('@react-navigation/native', () => {
    const actualNav = jest.requireActual('@react-navigation/native');
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: jest.fn(),
        }),
        useRoute: () => ({
            params: {
                username: 'testuser',
                userId: 'testuser-id'
            }
        })
    };
});

// Mock expo-image-picker
jest.mock('expo-image-picker', () => ({
    launchImageLibraryAsync: jest.fn(() => ({
        canceled: false,
        assets: [{ uri: 'test-uri' }]
    }))
}));

// Mock supabase
jest.mock('../../lib/supabase', () => ({
    supabase: {
        storage: {
            from: () => ({
                upload: jest.fn(() => ({ error: null })),
                getPublicUrl: jest.fn(() => ({ data: { publicUrl: 'test-url' } }))
            })
        },
        from: () => ({
            update: jest.fn(() => ({ error: null }))
        }),
        auth: {
            getUser: jest.fn(() => ({ data: { user: { id: 'testuser-id' } } }))
        }
    }
}));

describe('ProfilePictureScreen', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<ProfilePictureScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contains the expected text elements', () => {
        const component = renderer.create(<ProfilePictureScreen />);
        const instance = component.root;

        expect(instance.findByProps({ testID: 'titleText' }).props.children).toBe('Add a Profile Picture');
        expect(instance.findByProps({ testID: 'subtitleText' }).props.children).toBe('Customize your profile by adding a Profile Picture!');
    });

    it('contains the profile picture button', () => {
        const component = renderer.create(<ProfilePictureScreen />);
        const instance = component.root;

        expect(instance.findByProps({ testID: 'profilePictureButton' })).toBeTruthy();
    });

    it('contains the Next button', () => {
        const component = renderer.create(<ProfilePictureScreen />);
        const instance = component.root;

        expect(instance.findByProps({ testID: 'nextButton' })).toBeTruthy();
    });

    it('contains the Skip button', () => {
        const component = renderer.create(<ProfilePictureScreen />);
        const instance = component.root;

        expect(instance.findByProps({ testID: 'skipButton' })).toBeTruthy();
    });
});
