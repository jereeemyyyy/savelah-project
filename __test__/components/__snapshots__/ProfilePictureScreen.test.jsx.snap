// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePictureScreen renders correctly 1`] = `
<View
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "backgroundColor": "#1f2937",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#6366f1",
          "borderBottomLeftRadius": 16,
          "borderTopRightRadius": 16,
          "marginLeft": 16,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight="20"
        bbWidth="20"
        color="white"
        data-slot="icon"
        fill="currentColor"
        focusable={false}
        height="20"
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 20,
              "width": 20,
            },
          ]
        }
        tintColor="white"
        vbHeight={24}
        vbWidth={24}
        width="20"
      >
        <RNSVGGroup
          fill={
            {
              "type": 2,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            clipRule={0}
            d="M11.03 3.97a.75.75 0 0 1 0 1.06l-6.22 6.22H21a.75.75 0 0 1 0 1.5H4.81l6.22 6.22a.75.75 0 1 1-1.06 1.06l-7.5-7.5a.75.75 0 0 1 0-1.06l7.5-7.5a.75.75 0 0 1 1.06 0Z"
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            fillRule={0}
            propList={
              [
                "fillRule",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <Text
    style={
      [
        {
          "fontSize": 36,
          "lineHeight": 40,
        },
        {
          "fontWeight": "700",
        },
        {
          "color": "#fff",
        },
        {
          "marginBottom": 16,
          "marginLeft": 16,
          "marginRight": 16,
          "marginTop": 16,
        },
      ]
    }
    testID="titleText"
  >
    Add a Profile Picture
  </Text>
  <Text
    style={
      [
        {
          "fontSize": 18,
          "lineHeight": 28,
        },
        {
          "color": "#9ca3af",
        },
        {
          "marginLeft": 16,
          "marginRight": 16,
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
        },
      ]
    }
    testID="subtitleText"
  >
    Customize your profile by adding a Profile Picture!
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "marginBottom": 16,
        "marginTop": 16,
        "opacity": 1,
      }
    }
    testID="profilePictureButton"
  >
    <View
      style={
        [
          {
            "width": 144,
          },
          {
            "height": 144,
          },
          {
            "backgroundColor": "#e5e7eb",
          },
          {
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "alignItems": "center",
        },
        {
          "marginBottom": 16,
          "marginLeft": 16,
          "marginRight": 16,
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#6366f1",
          "borderBottomLeftRadius": 4,
          "borderBottomRightRadius": 4,
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
          "marginBottom": 16,
          "opacity": 1,
          "paddingBottom": 12,
          "paddingTop": 12,
          "width": 80,
        }
      }
      testID="nextButton"
    >
      <Text
        style={
          [
            {
              "color": "#fff",
            },
          ]
        }
      >
        Next
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#6b7280",
          "borderBottomLeftRadius": 4,
          "borderBottomRightRadius": 4,
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
          "opacity": 1,
          "paddingBottom": 12,
          "paddingTop": 12,
          "width": 80,
        }
      }
      testID="skipButton"
    >
      <Text
        style={
          [
            {
              "color": "#fff",
            },
          ]
        }
      >
        Skip
      </Text>
    </View>
  </View>
</View>
`;
